crates/flox-rust-sdk/src/models/environment/path_environment.rs

```diff
@@ -398,83 +403,66 @@ impl PathEnvironment<Original> {
+            Err(EnvironmentError2::EnvNotFound)?;
+        }
+
+        let env_path = env_dir
             .canonicalize()
             .map_err(EnvironmentError2::EnvCanonicalize)?;
+
+        // TODO: replace with checks for manifest file once we transition
         if !env_path.join("flake.nix").exists() {
             Err(EnvironmentError2::DirectoryNotAnEnv)?
         }
 
-        let name = EnvironmentName::from_str(&env_dir_entry.file_name().to_string_lossy())?;
-
         Ok(PathEnvironment {
             path: env_path,
-            name,
+            pointer,
             temp_dir: temp_dir.as_ref().to_path_buf(),
             state: Original,
         })
     }
 
-    /// Find the closest `.flox` starting with `current_dir`
-    /// and looking up ancestor directories until `/`
-    pub fn discover(
-        current_dir: impl AsRef<Path>,
-        temp_dir: impl AsRef<Path>,
-    ) -> Result<Option<Self>, EnvironmentError2> {
-        let maybe_dot_flox = current_dir
-            .as_ref()
-            .ancestors()
-            .find(|ancestor| ancestor.join(".flox").exists());
-
-        let with_dot_flox = if let Some(with_dot_flox) = maybe_dot_flox {
-            with_dot_flox
-        } else {
-            return Ok(None);
-        };
-
-        Some(Self::open(with_dot_flox, temp_dir)).transpose()
-    }
-
     /// Create a new env in a `.flox` directory within a specific path or open it if it exists.
     ///
     /// The method creates or opens a `.flox` directory _contained_ within `path`!
     pub fn init(
+        pointer: PathPointer,
         path: impl AsRef<Path>,
-        name: EnvironmentName,
         temp_dir: impl AsRef<Path>,
     ) -> Result<Self, EnvironmentError2> {
-        if Self::open(&path, &temp_dir).is_ok() {
-            Err(EnvironmentError2::EnvironmentExists)?;
+        match EnvironmentPointer::open(path.as_ref()) {
+            Err(EnvironmentError2::DirectoryNotAnEnv) => {},
+            Err(e) => Err(e)?,
+            Ok(_) => Err(EnvironmentError2::EnvironmentExists)?,
         }
 
-        let env_dir = path.as_ref().join(".flox").join(name.as_ref());
+        let dot_flox_path = path.as_ref().join(DOT_FLOX);
 
+        let env_dir = dot_flox_path.join(ENVIRONMENT_DIR_NAME);
         std::fs::create_dir_all(&env_dir).map_err(EnvironmentError2::InitEnv)?;
 
+        let pointer_content =
+            serde_json::to_string_pretty(&pointer).map_err(EnvironmentError2::SerializeEnvJson)?;
+
         copy_dir_recursive(&env!("FLOX_ENV_TEMPLATE"), &env_dir, false)
             .map_err(EnvironmentError2::InitEnv)?;
 
-        Self::open(path, temp_dir)
+        if let Err(e) = fs::write(
+            dot_flox_path.join(ENVIRONMENT_POINTER_FILENAME),
+            pointer_content,
+        ) {
+            fs::remove_dir_all(env_dir).map_err(EnvironmentError2::InitEnv)?;
+            Err(EnvironmentError2::WriteEnvJson(e))?;
+        }
+
+        Self::open(pointer, dot_flox_path, temp_dir)
```

Comment:
suggestion nonblocking: we don't need to open the environment as we only use this in `flox init` which doesn't actually do anything with the `PathEnvironment`

Reply:
My thoughts on this were:
- `void` methods make me uncomfortable (feel free to ignore)
- its a common api having a methods that creates an abstraction also return a handle to it
- the spec keeps it open to automatically create/init an environment if no env is found for install/edit so we might actually do something with it.
- its a useful validation step